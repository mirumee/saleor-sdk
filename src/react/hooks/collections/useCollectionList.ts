import { useApolloClient } from "react-apollo";
import { useEffect, useRef, useState } from "react";
import ApolloClient from "apollo-client";
import {
  CollectionList as CollectionListQuery,
  CollectionListVariables,
} from "../../../queries/gqlTypes/CollectionList";
import { BaseCollection } from "../../../fragments/gqlTypes/BaseCollection";
import { ListParameters } from "../../../api/types";
import { collections } from "../../../queries/collections";
import { CollectionList } from "../../../api/collections/CollectionList";

function query(client: ApolloClient<any>) {
  return (variables: CollectionListVariables) =>
    client.query<CollectionListQuery, CollectionListVariables>({
      query: collections,
      variables,
    });
}

function useCollectionList(params: ListParameters) {
  const client = useApolloClient();
  const [data, setData] = useState<BaseCollection[] | undefined>(undefined);
  const list = useRef<CollectionList>();
  const willMount = useRef(true);

  async function initList() {
    list.current = new CollectionList(query(client), params.count);
    await list.current.init({
      after: params.endCursor,
      first: params.count,
    });
    setData(list.current.data);
  }

  useEffect(() => {
    if (!willMount.current) {
      initList();
    }
  }, [params.count, params.endCursor]);

  if (willMount.current) {
    initList();
    willMount.current = false;
  }

  const next = async () => {
    const nextData = await list.current?.next();
    setData(nextData);
  };

  return {
    current: list.current?.current,
    data,
    loading: list.current?.loading,
    next,
    pageInfo: list.current?.pageInfo,
  };
}

export default useCollectionList;
