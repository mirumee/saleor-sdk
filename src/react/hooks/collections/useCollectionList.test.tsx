import { act, renderHook } from "@testing-library/react-hooks";
import React from "react";
import { ApolloProvider } from "react-apollo";
import setupAPI from "../../../../testUtils/api";
import useCollectionList from "./useCollectionList";

const { client } = setupAPI();

const wrapper: React.FC = ({ children }) => (
  <ApolloProvider client={client}>{children}</ApolloProvider>
);

describe("useCollection", () => {
  it("can fetch collections", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          count: 20,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can fetch next page", async () => {
    const { result } = renderHook(
      () =>
        useCollectionList({
          count: 1,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);

    const nextPromise = act(result.current.next);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(true);

    await act(() => nextPromise);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });
});
