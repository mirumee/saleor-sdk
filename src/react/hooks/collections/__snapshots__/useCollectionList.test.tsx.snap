// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCollection can fetch collections 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjoy",
    "name": "Winter sale",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "winter-sale",
  },
]
`;

exports[`useCollection can fetch next page 1`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
]
`;

exports[`useCollection can fetch next page 2`] = `
Array [
  Object {
    "__typename": "Collection",
    "id": "Q29sbGVjdGlvbjox",
    "name": "Summer collection",
    "seoDescription": null,
    "seoTitle": null,
    "slug": "summer-collection",
  },
]
`;
